hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestCoalescingPartitioner.java

```diff
@@ -0,0 +1,111 @@
+ */
+
+package org.apache.hudi.client;
+
+import org.apache.hudi.common.data.HoodieData;
+import org.apache.hudi.data.HoodieJavaRDD;
+import org.apache.hudi.testutils.HoodieClientTestBase;
+
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.api.java.function.PairFunction;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import scala.Tuple2;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TestCoalescingPartitioner extends HoodieClientTestBase {
+
+  @Test
+  public void simpleCoalescingPartitionerTest() {
+    int numPartitions = 100;
+    HoodieData<Integer> rddData = HoodieJavaRDD.of(jsc.parallelize(
+        IntStream.rangeClosed(0, 100).boxed().collect(Collectors.toList()), numPartitions));
+
+    // 100 keys spread across 10 partitions.
+    CoalescingPartitioner coalescingPartitioner = new CoalescingPartitioner(10);
+    assertEquals(10, coalescingPartitioner.numPartitions());
+    rddData.collectAsList().forEach(entry -> {
+      assertEquals(entry.hashCode() % 10, coalescingPartitioner.getPartition(entry));
+    });
+
+    // 1 partition
+    CoalescingPartitioner coalescingPartitioner1 = new CoalescingPartitioner(1);
+    assertEquals(1, coalescingPartitioner1.numPartitions());
+    rddData.collectAsList().forEach(entry -> {
+      assertEquals(0, coalescingPartitioner1.getPartition(entry));
+    });
+
+    // empty rdd
+    rddData = HoodieJavaRDD.of(jsc.emptyRDD());
+    CoalescingPartitioner coalescingPartitioner2 = new CoalescingPartitioner(1);
+    assertEquals(1, coalescingPartitioner2.numPartitions());
+    rddData.collectAsList().forEach(entry -> {
+      // since there is only one partition, any getPartition will return just the same partition index
+      assertEquals(0, coalescingPartitioner2.getPartition(entry));
+    });
+  }
+
+  private static Stream<Arguments> coalesceTestArgs() {
+    return Arrays.stream(new Object[][] {
+        {100, 1},
+        {1, 1},
+        {1000, 10},
+        {100, 7},
+        {10, 2}
+    }).map(Arguments::of);
+  }
+
+  @ParameterizedTest
+  @MethodSource("coalesceTestArgs")
+  public void testCoalescingPartitionerWithRDD(int inputNumPartitions, int targetPartitions) {
+    List<Integer> inputData = IntStream.rangeClosed(0, inputNumPartitions).boxed().collect(Collectors.toList());
+    JavaRDD<Integer> data = jsc.parallelize(inputData, inputNumPartitions);
+    JavaRDD<Integer> coalescedData = data.mapToPair(new PairFunc()).partitionBy(new CoalescingPartitioner(targetPartitions)).map(new MapFunc());
+
+    assertEquals(targetPartitions, coalescedData.getNumPartitions());
```

Comment:
We should check for a relatively even spread amongst the partitions. I wrote some local code to validate that the partitioner is unfortunately writing all data to the same spark partition:

```
List<Pair<Integer, Integer>> results = coalescedData.mapPartitionsWithIndex((i, rows) -> {
      int count = 0;
      while (rows.hasNext()) {
        rows.next();
        count++;
      }
      return Collections.singletonList(Pair.of(i, count)).iterator();
    }, true).collect();

```